### PR TITLE
Plans: add plan matching function for yearly/monthly plans

### PR DESCRIPTION
**Pull request by sirreal**
`isPlanMatch` helps to identify plans regardless of their interval.

To test:
* ensure tests pass `npm run test-client`
* Ensure plan features continue to render correctly monthly/yearly (make sure the import/export I moved works correctly): http://calypso.localhost:3000/plans/$jetpackSite

### CODE REVIEW

File: client/lib/plans/test/index.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+		];
+		matchingPlans.forEach(
+			( [ slugA, slugB ] ) => expect( isPlanMatch( slugA, slugB ) ).to.be.true
+		);
+	} );
+
+	it( 'should return false for non-matching plans', () => {
+		expect( isPlanMatch( PLAN_JETPACK_PREMIUM, PLAN_JETPACK_PERSONAL ) ).to.be.false;
+	} );
+} );
```
Comment (tyxla) [inline client/lib/plans/test/index.js]: Maybe it's worth to include a couple of tests with non-Jetpack plans and even non-plan products.

File: client/lib/plans/test/index.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	PLAN_JETPACK_BUSINESS_MONTHLY,
+	PLAN_JETPACK_PERSONAL,
+	PLAN_JETPACK_PERSONAL_MONTHLY,
+	PLAN_JETPACK_PREMIUM,
+	PLAN_JETPACK_PREMIUM_MONTHLY,
+} from '../constants';
+
+describe( 'isPlanMatch', () => {
+	it( 'should return true for identical plans', () => {
+		expect( isPlanMatch( PLAN_JETPACK_BUSINESS, PLAN_JETPACK_BUSINESS ) ).to.be.true;
```
Comment (tyxla) [inline client/lib/plans/test/index.js]: Do we also want to test that case for all possible plans that we're covering?

File: client/lib/plans/test/index.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		const matchingPlans = [
+			[ PLAN_JETPACK_BUSINESS, PLAN_JETPACK_BUSINESS_MONTHLY ],
+			[ PLAN_JETPACK_PERSONAL, PLAN_JETPACK_PERSONAL_MONTHLY ],
+			[ PLAN_JETPACK_PREMIUM, PLAN_JETPACK_PREMIUM_MONTHLY ],
+			[ PLAN_JETPACK_BUSINESS_MONTHLY, PLAN_JETPACK_BUSINESS ],
+			[ PLAN_JETPACK_PERSONAL_MONTHLY, PLAN_JETPACK_PERSONAL ],
+			[ PLAN_JETPACK_PREMIUM_MONTHLY, PLAN_JETPACK_PREMIUM ],
+		];
+		matchingPlans.forEach(
+			( [ slugA, slugB ] ) => expect( isPlanMatch( slugA, slugB ) ).to.be.true
```
Comment (tyxla) [inline client/lib/plans/test/index.js]: Another option for writing this one is to cut the `matchingPlans` in half, and just have two `expect` calls for each of the matching plans:

```
it( 'should return true for matching plans', () => {
	const matchingPlans = [
		[ PLAN_JETPACK_BUSINESS, PLAN_JETPACK_BUSINESS_MONTHLY ],
		[ PLAN_JETPACK_PERSONAL, PLAN_JETPACK_PERSONAL_MONTHLY ],
		[ PLAN_JETPACK_PREMIUM, PLAN_JETPACK_PREMIUM_MONTHLY ],
	];
	matchingPlans.forEach(
		( [ slugA, slugB ] ) => {
			expect( isPlanMatch( slugA, slugB ) ).to.be.true;
			expect( isPlanMatch( slugB, slugA ) ).to.be.true;
		}
	);
} );
```

Less code, probably a bit more complexity. Your call 😉

File: client/lib/plans/index.js [hunk 4]
```diff
@@ -143,6 +149,36 @@ export function filterPlansBySiteAndProps( plans, site, hideFreePlan, intervalTy
...
+/**
+ * Returns true if plans match regardeless of their interval.
+ * For example (fake plans):
+ *     isPlanMatch( 'proYearly', 'proYearly' ) => true
+ *     isPlanMatch( 'proYearly', 'proMonthly' ) => true
+ *     isPlanMatch( 'proYearly', 'personalYearly' ) => false
+ *
+ * @param  {String}  planSlugA One of the plan slugs to compare
+ * @param  {String}  planSlugB One of the plan slugs to compare
+ * @return {Boolean}           Whether the plans match
```
Comment (sirreal) [inline client/lib/plans/index.js]: In the `@return`? I think the description is pretty clear.

File: client/lib/plans/index.js [hunk 5]
```diff
@@ -143,6 +149,36 @@ export function filterPlansBySiteAndProps( plans, site, hideFreePlan, intervalTy
...
+ */
+export function isPlanMatch( planSlugA, planSlugB ) {
```
Comment (sirreal) [inline client/lib/plans/index.js]: You mean `isPlanMatch`? It doesn't feel right to me either, I was hoping my reviewers would have good suggestions 😅 

I've been thinking about it a fair amount and really don't know how to express this without getting overly verbose with something like `isPlanMatchIgnoringIntervals`. Any ideas?

Comment (seear) [inline client/lib/plans/index.js]: `planLevelsMatch()`,
`matchIgnoringIntervals`()`

Comment (tyxla) [inline client/lib/plans/index.js]: Good suggestions @seear, mine initially were:

* `planTypesMatch()`
* `plansHaveMatchingTypes()` - any WordPress theme/plugin developer would appreciate having the word `have` in the function name 😆 

But I'd leave that to native speakers. As @sirreal mentioned, we don't want to get too verbose, otherwise something like `arePassedPlanSlugsMatchingTheirPlanTypesIgnoringTheIntervals()` would be great 😉

File: client/lib/plans/index.js [hunk 6]
```diff
@@ -143,6 +149,36 @@ export function filterPlansBySiteAndProps( plans, site, hideFreePlan, intervalTy
...
+	return planSlugA === planSlugB ||
```
Comment (seear) [inline client/lib/plans/index.js]: I'm late to this, so feel free to ignore, but another way to do it is have a `convertToYearlyPlan()` (or monthly) func that returns yearly plans unchanged, then this becomes simply:

`return convertToYearly( planSlugA) === convertToYearly( planSlugB);`

Comment (tyxla) [inline client/lib/plans/index.js]: @seear is right here, if we have that one, we could use it in some other locations to avoid duplicating plans logic (if I remember correctly @sirreal, there were several cases in your plans refactor PRs where this could have been useful).

File: client/lib/plans/index.js (hunk 4 above)
Comment (tyxla) [inline client/lib/plans/index.js]: Sorry, I should have clarified a bit more. I mean that two plan could match by their type, or by their interval, right. And here we don't specify which criteria they they are matched by. I'd rephrase this as "Whether the plan levels match" or "Whether the plan types match", or even add "regardless of the plan interval". Maybe I'm just missing a JSDoc, which could contain all of that.

File: client/lib/plans/index.js (hunk 5 above)
Comment (sirreal) [inline client/lib/plans/index.js]: Went for `planLevelsMatch`, thanks both.

File: client/lib/plans/index.js (hunk 6 above)
Comment (sirreal) [inline client/lib/plans/index.js]: I agree that the following is cleaner and more elegant:

```
return convertToYearly( planSlugA) === convertToYearly( planSlugB);
```

Unfortunately, the way `getYearlyPlanByMonthly` and `getMonthlyPlanByYearly` have been implemented, they return the corresponding version version or `""`. This makes it easy to get false positives and indeed the tests fail when using the suggested implementation. By relying on testing a real slug (not `""`) against a converted slug (which may be `""`), we reduce the chance of these bad results.
